File: android/src/main/java/com/unact/yandexmapkit/YandexPolylineController.java
package com.unact.yandexmapkit;

import com.yandex.mapkit.map.MapObjectCollection;
import com.yandex.mapkit.map.PolylineMapObject;

import java.lang.ref.WeakReference;
import java.util.Map;

public class YandexPolylineController extends YandexMapObjectController {
  public final PolylineMapObject polyline;
  private final YandexMapObjectTapListener tapListener;
  @SuppressWarnings({"UnusedDeclaration", "FieldCanBeLocal"})
  private final WeakReference<YandexMapController> controller;
  public final String id;

  public YandexPolylineController(
    MapObjectCollection parent,
    Map<String, Object> params,
    WeakReference<YandexMapController> controller
  ) {
    PolylineMapObject polyline = parent.addPolyline(Utils.polylineFromJson(params));

    this.polyline = polyline;
    this.id = (String) params.get("id");
    this.controller = controller;
    this.tapListener = new YandexMapObjectTapListener(id, controller);

    polyline.setUserData(this.id);
    polyline.addTapListener(tapListener);
    update(params);
  }

  @SuppressWarnings({"unchecked", "ConstantConditions"})
  public void update(Map<String, Object> params) {
    polyline.setGeometry(Utils.polylineFromJson(params));
    polyline.setGeodesic((Boolean) params.get("isGeodesic"));
    polyline.setZIndex(((Double) params.get("zIndex")).floatValue());
    polyline.setVisible((Boolean) params.get("isVisible"));
    polyline.setOutlineColor(((Number) params.get("outlineColor")).intValue());
    polyline.setOutlineWidth(((Double) params.get("outlineWidth")).floatValue());
    polyline.setStrokeColor(((Number) params.get("strokeColor")).intValue());
    polyline.setStrokeWidth(((Double) params.get("strokeWidth")).floatValue());
    polyline.setDashLength(((Double) params.get("dashLength")).floatValue());
    polyline.setDashOffset(((Double) params.get("dashOffset")).floatValue());
    polyline.setGapLength(((Double) params.get("gapLength")).floatValue());
  }

  public void remove() {
    polyline.getParent().remove(polyline);
  }
}

File: ios/Classes/YandexPolylineController.swift
import YandexMapsMobile

class YandexPolylineController: NSObject, YandexMapObjectController, YMKMapObjectTapListener {
  public let polyline: YMKPolylineMapObject
  private var consumeTapEvents: Bool = false
  public unowned var controller: YandexMapController
  public let id: String

  public required init(
    parent: YMKMapObjectCollection,
    params: [String: Any],
    controller: YandexMapController
  ) {
    let polyline = parent.addPolyline(with: Utils.polylineFromJson(params))

    self.polyline = polyline
    self.id = params["id"] as! String
    self.controller = controller

    super.init()

    polyline.userData = self.id
    polyline.addTapListener(with: self)
    update(params)
  }

  public func update(_ params: [String: Any]) {
    polyline.geometry = Utils.polylineFromJson(params)
    polyline.zIndex = (params["zIndex"] as! NSNumber).floatValue
//    polyline.isGeodesic = (params["isGeodesic"] as! NSNumber).boolValue
    polyline.isVisible = (params["isVisible"] as! NSNumber).boolValue
//    polyline.strokeColor = Utils.uiColor(fromInt: (params["strokeColor"] as! NSNumber).int64Value)
    polyline.outlineColor = Utils.uiColor(fromInt: (params["outlineColor"] as! NSNumber).int64Value)
    polyline.outlineWidth = (params["outlineWidth"] as! NSNumber).floatValue
    polyline.strokeWidth = (params["strokeWidth"] as! NSNumber).floatValue
    polyline.dashLength = (params["dashLength"] as! NSNumber).floatValue
    polyline.dashOffset = (params["dashOffset"] as! NSNumber).floatValue
    polyline.gapLength = (params["gapLength"] as! NSNumber).floatValue

    consumeTapEvents = (params["consumeTapEvents"] as! NSNumber).boolValue
  }

  public func remove() {
    polyline.parent.remove(with: polyline)
  }

  func onMapObjectTap(with mapObject: YMKMapObject, point: YMKPoint) -> Bool {
    controller.mapObjectTap(id: id, point: point)

    return consumeTapEvents
  }
}
